# Part 01 - Immediate Applications

Python strikes a good balance between simplicity and its capabilities. Upon [installation](https://www.python.org/downloads/), you can already program some immediate applications by applying some arithmetic _expressions_.

## Arithmetic

Some simple addition:

In [13]:
56 + 65

121

Subtraction:

In [15]:
65 - 56

9

Division:

In [4]:
9 / 3

3.0

How about:

In [18]:
2 ** 4 + 3 + 2 + 1

22

The double asterisks `**` tells Python to perform exponentiation for a _base_ of `2` and an _exponent/power_ of `4`. The rest of the expression evaluates follow the natural order from left to right and eventually resulted in `22`.

In [19]:
2 ** (4 + 3 + 2 + 1)

1024

Notice the parentheses used around the exponent/power. That is a direct convention borrowed from mainstream mathematics to alterate evaluation precedences.

Python has [a table of built-in expression evaluation precedence](https://docs.python.org/3.8/reference/expressions.html#operator-precedence) which largely inherits from relevant-domain mainstreams.

In this case, the parentheses ensure that Python understands it as taking a base `2` and an exponent of `4 + 3 + 2 + 1`, which means that the evaluation process would first yield the exponent to be `10`, and then `2 ** 10` which finally results in `1024`. When in doubt of evaluation order, apply parentheses liberally to annotate your true will.

Back to division, with some "real-world" application, such as dividing 10 dollars by 3: 

In [21]:
10.00 / 3

3.3333333333333335

Notice the oddity in the output value `3.3333333333333335`?

It is known as a floating-point arithmetic error or round-off error. It is a fundamental trade-off made back in the process of digitization (where real numbers are only "emulated" by their binary representations). So by design, this type of error cannot be eliminated in digitally based systems but can only be managed by taking trade-offs between ranges and precisions. When it's not managed well, [it resulted in matters of life and death](https://en.wikipedia.org/wiki/Round-off_error#Real_world_example:_Patriot_missile_failure_due_to_magnification_of_roundoff_error).

Floating-point arithmetic error is the reason why precision-sensitive data is stored and calculated in fixed-point numbers -- or integers with an implicit definition of the fraction scale it represents. For instance, if we want to perform the same division, instead of `10.00 / 3`, we'll scale everything to cents (1/100 of a dollar):

In [22]:
1000 / 3

333.3333333333333

Now it's a bit better without the odd `5` at the end of the fraction. And if we choose a desired fractional precision and can afford to forego the remainder, let's say down to 1/10 of a cent (1/1000 of a dollar):

In [10]:
10000 // 3

3333

The `//` operator denotes specifically for an integer division, or a mathematical _floor_.

And if we want to know how much remainder we've foregone, we can use the _modulo_ operator `%` to do so:

In [11]:
10000 % 3

1

Which is 1/10 of a cent, or 1 mill in the US monetary system. I'm sure most people are okay with that, except for those who care about how some financial institutions have been making extra doughs legally and openly this way forever.

## Expressions

The above arithmetics are all expressions. And like every living natural language in the world, expressions follow syntactical and semantic rules.

In [23]:
-1000

-1000

In [24]:
+1000

1000

In [25]:
1000-

SyntaxError: invalid syntax (<ipython-input-25-be6920bdb29f>, line 1)

The first two expressions `-1000` and `+1000` resolved as expected (where their signs are enforced). But the third expression `1000-` resulted in a error (`SyntaxError: invalid syntax`).

But don't panic. Errors are okay and Python usually provides good enough context for you to know what kind of error you have encountered so you can [search around](https://duckduckgo.com/?q=Python+SyntaxError%3A+invalid+syntax&t=ffab&ia=web) for workarounds or solutions.

In this particular case, the error indicate that there's a missing operand after the `-` operator. So let's fix it with one:

In [26]:
1000-1

999

Numbers and arithemetics are not the only possible expressions you can make with Python.

Writing a string of characters is also perfectly valid, such as:

In [32]:
'MAKE PUPPIES GREAT AGAIN!'

'MAKE PUPPIES GREAT AGAIN!'

In [31]:
"alphanumerics are 'a-z' and '0-9'"

"alphanumerics are 'a-z' and '0-9'"

Both of above are examples of Python string (`str`) type. They are usually expressed through single `'` or double `"` quotations around the intended characters.

In [36]:
'''Anyone who has never made a mistake has never tried anything new.

	- Albert Einstein, Apparently'''

'Anyone who has never made a mistake has never tried anything new.\n\n\t- Albert Einstein, Apparently'

The triple-quotes (`'''` or equally valid `"""`) surrounded `string`s are special since they preserve special characters that represent newlines `\n` and tabs `\t`. They are not apparent to our eyes unless we `print` it out though:

In [37]:
print('''Anyone who has never made a mistake has never tried anything new.

	- Albert Einstein, Apparently''')

Anyone who has never made a mistake has never tried anything new.

	- Albert Einstein, Apparently


`print()` here is a Python built-in _function_. We'll cover functions in greater details later in the series.

For now, all we have to know is that between its parentheses, we can fill in any expressions we've learned so far:

In [38]:
print(56 + 65)

121


So what's the difference between `56 + 65` from before and `print(56 + 65)`? There a few, but most notably:

* `56 + 65` evaluation returns us `121` as a number type, or more specifically, an integer (`int`) type.
* `print(56 + 65)` evaluation returns us nothing (`NoneType`) type, while "printing" it out to our screen, or more technically `stdout` (standard output), one of the [standard streams](https://en.wikipedia.org/wiki/Standard_streams) modern operating systems implement that allow programs to communicate.

These can be easily verified through the use of another built-in function `type()`:

In [39]:
type(56 + 65)

int

In [40]:
type(print(56 + 65))

121


NoneType

Together with characters string (`str`) type and floating-point numbers (`float`) type from above, these cover the 4 most primitive data types used in Python applications:

In [41]:
type(10.00)

float

In [42]:
type('MAKE PUPPIES GREAT AGAIN!')

str

To effectively utilize these data types, we'll need to follow some well-defined rules, or we'll get errors:

In [1]:
2 + '1'

TypeError: unsupported operand type(s) for +: 'int' and 'str'

The above error (a `TypeError`) state that the error happens within the context of unsupported `+` operation between two types `int` and `str`.

Similarly:

In [51]:
'65 + 56 equals ' + 121

TypeError: can only concatenate str (not "int") to str

Interestingly, `+` operation can be applied to them safely when both operands are of the same/similar type:

In [52]:
'65 + 56 equals ' + '121'

'65 + 56 equals 121'

In [50]:
2 + 1.0

3.0

We can observe from above that `+` between two numbers result in an addition, while applied to strings as concatenation. The logical explanation is that strings are a _collection_ of characters, thus "addition" of two collections would result in them concatenate with each other.

Another apparent mathematical operator that applies to strings to a (slightly) different effect is the multiplication operator `*`. With strings, it repeats the existing collection of characters by a number of specified times:

In [55]:
print('I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM\n' * 10)

I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM
I WILL NOT YELL “FIRE” IN A CROWDED CLASSROOM



The ["chalkboard gag"](https://simpsons.fandom.com/wiki/Chalkboard_gag) wouldn't be such a meme if Bart Simpsons knew Python.

## Variables

So far we've applied Python through some of the immediate ways but not so "reusable". _Variables_ are one of the first enablers for "reusability" in programs, since they are responsible for temporary storage of values, and act as a layer of abstraction, as in its mathematical concept.

Take a formula such as:
![variable](https://i.imgur.com/IGTidUL.png)

And turn it into Python code:

In [56]:
a * x ** 2 + b * x + c

NameError: name 'x' is not defined

The formula is abstract, which means it doesn't hold specific values in any of its 4 variables. By naively translate the formula into Python code, we encountered an error (`NameError`) because Python expressions would need materialized values to be evaluated.

This means that we'd need to _define_ the said variables, and also _assign_ them with meaningful values for the formula expression to work.

In [3]:
# define necessary variables
a = 1
b = 2
c = 3
x = 5

a * x ** 2 + b * x + c

38

With the same formula, but this time with its variables defined and assigned through an _assignment statement_ `=`, it worked as expected.

In [4]:
# re-assign some variables
a = 5
x = 1

a * x ** 2 + b * x + c

10

What happened above was that we successfully reused the same formula, after re-assigning two of the variables.

This signifies one of the biggest characteristics of the Python programming language -- variables can be re-assigned and they can even be re-assigned to different types:

In [8]:
a = 'not a number anymore'
print(a)

not a number anymore


Here variable `a` was re-assigned with a string value instead of a number (or more precisely, an integer) and Python interpreted it well. But after this, the same forumla we've been reusing would fail:

In [7]:
a * x ** 2 + b * x + c

TypeError: can only concatenate str (not "int") to str

As expected, and seen before.

Variables and _types_ are highly correlated concepts in most programming languages. Python takes a dynamic approach and thus a variable doesn't need to be defined with a fixed type before a value is assigned to it.

In [59]:
v1 = 1
v1 = 2.0
v1 = 'a phrase'

type(v1)

str

The same variable `v1` went through 3 assignment statements of 3 different types, and its eventual type corresponded to its last assignment.

In fact, variables can also hold functions as its value, since functions are one of the `Callable` types in Python.

In [63]:
vf = print

type(vf)

builtin_function_or_method

Variables may be loose in terms of what type of data it can hold and be reassigned, but there are strict rules on how they can be named.

In Python, variable names must follow the following rules:
* Only characters, numbers and `_` are allowed, assembed in one word without empty spaces or newlines
* Cannot start with a number
* Cannot include special characters or expression operators

Some counter-examples:

In [65]:
4a = 1  # start with a number

SyntaxError: invalid syntax (<ipython-input-65-843f32448d8a>, line 1)

In [66]:
\b = 'this is b'  # start with a special character

SyntaxError: unexpected character after line continuation character (<ipython-input-66-4df936ec44a6>, line 1)

In [70]:
-d = 5  # start with an operator

SyntaxError: cannot assign to operator (<ipython-input-70-62d6c3855640>, line 1)

In [68]:
broken word = 'who let the dogs out'

SyntaxError: invalid syntax (<ipython-input-68-ce4905fe9f93>, line 1)

And some valid examples:

In [71]:
a4 = 1

In [72]:
b = 'this is b'

In [73]:
d = -5

In [74]:
broken_word = 'who let the dogs out'

Also observe that above assignment do not have any output. That's because an assignment statement by definition does not evaluate but only persist the value from the right hand side of `=`. However, the right hand side of the statement _can_ be any valid expression since expressions result in an eventual value.

In [75]:
evaluated = 12 + 21

And assigned variables can be used in other expressions:

In [76]:
print(evaluated - 33)

0
